### ISSUE TITLE
reuse of access tokens

### ISSUE THREAD

**Issue by pvdbosch** [belgif/rest-security#10]
https://github.com/belgif/rest-security/pull/4 forbids reuse of a token:

> The resource server SHOULD rely on the unique identifier provided by the "jti" (JWT identifier) claim to enforce the non-reuse of access tokens.
> Indeed, access tokens MUST be short-lived (lifetime should not exceed a few minutes) and used at most once.

What is the rationale behind this? Reuse of access tokens seems to be a common industry practice.

Some considerations I can think of:
* blocking reuse of tokens makes it harder for an attacker to abuse intercepted tokens. TLS should prevent this however. 
* allowing reuse is more efficient because it doesn't require a client to call the authorization server on each request. When using token introspect, the introspect response could be cached as well by the resource server.
* no token reuse for authorization code flow, probably means additional redirects for web applications (?), disrupting the user experience

I'll put this point on the agenda with SocSec/Smals security experts

**Comment by ghost** [comment]
@pvdbosch,

That's a good question.

JWTs are self-contained and thus - by design - remain valid for their defined lifetime and cannot be invalidated/revoked once issued, since no validation request is made to the issuing authorization server (unless each and every token is validated on the authorization server, which is however quite silly and counterproductive in most - but not all - scenarios).
This makes them very vulnerable to replay attacks.

While TLS protects against MITM, making it difficult to intercept valid tokens for replay attacks, even in ~~2021~~ 2022, it's unfortunately not that uncommon to see things like unencrypted "internal traffic" or token expiration set to a ridiculously high value by default (usually 1 hour if not more, since the industry is also plagued by bad practices).

Therefore, it is obvious that additional countermeasures must be put in place (as described in the guide).
Fortunately, smart people have already thought of this scenario, and section 4.1.7 of RFC 7519, introduces the jti (JWT ID) claim, i.e., a unique identifier for the token, specifically to prevent the JWT from being replayed.

If we add to this the fact that we impose the use of the TLS protocol in the guide and that we recommend a fairly short expiration time (from 30 seconds to a few minutes maximum, with a tolerance of a few seconds - no more than 30 - in the code to account for possible clock skew), we end up with several defensive mechanisms in place to protect against replay attacks (defense-in-depth approach).

This approach has, of course, obvious drawbacks that must be clearly understood.

First, it means storing tokens that have already been used, thus a state. In my opinion, it is better to move the state to a (distributed) cache or key value store to keep most services stateless.
However, a client that stores an access token for an hour or so can also be considered "stateful." Moreover, that state is even riskier because there is a good chance that the token is not stored in a secure fashion.

Second, one might even consider this approach to be inefficient, as it implies that a new access token must be issued for each request, and thus additional calls on the authorization server.
Nevertheless, while the adverse impact on performance is undeniable, it remains somewhat limited. 
Indeed, in scenarios not involving a user, the client credentials grant should be used and, as per section 4.4.3 of RFC 6749, a refresh token should not be returned.
In scenarios involving a user, such as browser-based single page applications, the PKCE grant (extension to the authorization code grant) should be used. It involves access tokens and refresh tokens passed over the backchannel (never through the browser).
Since the refresh tokens must also be single-use, the number of requests to the authorization server is indeed doubled (the user experience is in principle not impacted as long as the authorization server is not overloaded and requests are processed in a timely manner).
However, if you are really so worried about this, I assume that you are at a point where performance is of utmost importance and cannot be further improved by, for instance, optimizing the code or having a service mesh for east-west communication.

In conclusion, I do believe that token reuse should be prohibited.
At the same time, I am also convinced that if you understand the consequences of your choices (ask yourself: are you willing to sacrifice security for the sake of performance, scalability and statelessness?) when defining your architecture, the disadvantages of this approach are absolutely negligible, especially if you consider the kind of data we are dealing with...

I trust this rather long explanation answers your question. In any case, we can discuss this topic in more detail at our next meeting.

**Comment by pvdbosch** [comment]
I agree there's an additional measure of security, though I'm not convinced yet that its costs (performance, complexity) justify its use always; maybe only for data classification level 3 or 4. Though in those cases, if you're using JWS on requests with an expiration time, it would also protect against replay attacks. 
For data classification levels 1 or 2, it could be recommended to limit the validity period of the access token.

I guess if the client could leak access tokens, an attack vector for stealing a refresh token still remains; regardless of the use of one-time access tokens.
